zerver/tests/test_signup.py

```diff
@@ -1470,6 +1467,38 @@ def test_signup_already_active(self) -> None:
         result = self.client_get(result.url)
         self.assert_in_response("You've already registered", result)
 
+    def test_signup_with_used_email(self) -> None:
+        """
+        Check if signing up with an email used in another realm succeeds.
+        """
+        email = self.example_email('hamlet')
+        password = "newpassword"
+        realm = get_realm('lear')
+        do_set_realm_property(realm, 'default_language', u"de")
+
+        result = self.client_post('/accounts/home/', {'email': email}, subdomain="lear")
+        self.assertEqual(result.status_code, 302)
+        self.assertTrue(result["Location"].endswith(
+            "/accounts/send_confirm/%s" % (email,)))
+        result = self.client_get(result["Location"], subdomain="lear")
+        self.assert_in_response("Check your email so we can get started.", result)
+
+        # Visit the confirmation link.
+        confirmation_url = self.get_confirmation_url_from_outbox(email)
+        result = self.client_get(confirmation_url, subdomain="lear")
+        self.assertEqual(result.status_code, 200)
+
+        # Pick a password and agree to the ToS.
+        result = self.submit_reg_form_for_user(email, password, subdomain="lear")
+        self.assertEqual(result.status_code, 302)
+
+        user_profile = get_user(self.example_email("hamlet"), realm)
+        self.assertEqual(user_profile.default_language, realm.default_language)
+        from django.core.mail import outbox
+        outbox.pop()
```

Comment:
Hmm, is this needed? Looks like some but not all of the other tests do it. The remaining line of this test doesn't look like it should care, and if it can affect other tests then that's a broader bug.

In `test_successful_resend_invitation` we do this and then resend the invitation, and there it makes more sense because it could matter for testing the resend.

Reply:
I believe Django's `outbox` is designed properly for tests, and the issue here is just with tests that send emails later.  I have mixed feelings on removing it.  In one sense, it's a good pattern for our registration tests to do this sort of clearing, because they are the sort of thing that are copied-and-extended, and it's a lot better to have a bit of extra code than to have a test behave incorrectly due to the bug one would get if you used outbox later and didn't have this code.

Reply:
I don't think it's a good pattern to have a bit of code that looks like it shouldn't matter but gets cargo-culted around -- it normalizes cargo-culting code, and expecting that code might sometimes be doing magical unpredictable things.

In the case of the outbox, I think it should be pretty clear if the later code cares about the outbox, and I think there are only a couple of tests where that's the case. So it should be a good follow-up to clean that out from all the places where it's been unnecessarily copied. For this branch, we can just skip propagating it further.